File: Patrones/Comportamiento/State/Python/MaquinaExpendedora.py
from EstadoConcreto import SinDinero


# Máquina Expendedora (Contexto)
class MaquinaExpendedora:
    def __init__(self):
        # Estado inicial: Sin Dinero
        self.__estado_actual = SinDinero()

    # Cambiar el estado actual
    def set_estado(self, nuevo_estado):
        self.__estado_actual = nuevo_estado

    # Métodos que delegan la operación al estado actual
    def insertar_dinero(self):
        self.__estado_actual.insertar_dinero(self)

    def seleccionar_producto(self):
        self.__estado_actual.seleccionar_producto(self)

    def dispensar(self):
        self.__estado_actual.dispensar(self)

File: Patrones/Comportamiento/State/Java/src/MaquinaExpendedora.java
public class MaquinaExpendedora {
    private Estado estadoActual;

    public MaquinaExpendedora() {
        // Estado inicial: Sin Dinero
        this.estadoActual = new SinDinero();
    }

    // Cambiar el estado actual
    public void setEstado(Estado nuevoEstado) {
        this.estadoActual = nuevoEstado;
    }

    // Métodos que delegan la operación al estado actual
    public void insertarDinero() {
        estadoActual.insertarDinero(this);
    }

    public void seleccionarProducto() {
        estadoActual.seleccionarProducto(this);
    }

    public void dispensar() {
        estadoActual.dispensar(this);
    }
}
